### ISSUE TITLE
using windows event sink log doesn't compile with UNICODE preprocessor

### ISSUE THREAD

**Issue by malidiab** [gabime/spdlog#1717]
in file spdlog/include/spdlog/sinks/win_eventlog_sink.h , the file use ::FormatMessage  which is defined as FormatMessageA for non UNICODE and FormatMessageW for UNICODE in windows headers . a default windows application will need to define UNICODE . Having UNICODE enabled gives compilation error below , because the function tries to use std::string instead of std::wstring with FormatMessage  .  The compilation error appears once you include the header file "win_eventlog_sink.h" 

> error C2664: 'DWORD FormatMessageW(DWORD,LPCVOID,DWORD,DWORD,LPWSTR,DWORD,va_list *)': cannot convert argument 5 from 'LPSTR' to 'LPWSTR'

` static std::string format(std::string const &user_message, DWORD error_code = GetLastError())
    {
        std::string system_message;

        LPSTR format_message_result{};
        auto format_message_succeeded =
            ::FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, nullptr,
                error_code, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)&format_message_result, 0, nullptr);`

**Comment by gabime** [comment]
So the fix would be to use ```FormatMessageA``` instead (assuming unicode will be not supported by spdlog for this) ?

**Comment by malidiab** [comment]
Yes it will fix it , it will stop the compilation error and will fix it for ANSI strings , but ... 
since every user of spdlog expects std::string to be UTF8 ( as I understand ) , they might use it using UTF8 encoded strings that has non English char , which will cause MS APIs like `FormatMessageA` to fail unexpectedly , since it expects either ANSI strings or 16BIT strings. 
based on what I understand , I think we have two options : 

1 - inside the function convert std::string to wide char on System APIs boundary ( before calling any MS APIs ) , and call the Unicode version of any windows API ( ends with W , like FormatMessageW ) 

2 - provide win_eventlog_sink_unicode" that use std::wstring and calls the unicode version of any windows api. ( I think this will require other changes in some other places ?? to convert the UTF8 strings received from the caller to wide , probably this is a bad solution )

**Comment by gabime** [comment]
> since every user of spdlog expects std::string to be UTF8

Actually, UTF8 is not always expected. They might be UTF8 if user defines SPDLOG_WCHAR_TO_UTF8_SUPPORT.

**Comment by malidiab** [comment]
yes you are right. but currently If I don't use that define SPDLOG_WCHAR_TO_UTF8_SUPPORT on windows  , I start to get compilation errors in the library it self ( that is a different problem ? )

`1>..\vcpkg_installed\x64-windows\include\spdlog\logger.h(93): error C2664: 'void spdlog::logger::log(spdlog::level::level_enum,spdlog::string_view_t)': cannot convert argument 3 from 'const wchar_t [74]' to 'spdlog::string_view_t'

1>..\vcpkg_installed\x64-windows\include\spdlog\spdlog.h(151): note: see reference to function template instantiation 'void spdlog::logger::info<wchar_t[74],utility::string_t>(const FormatString (&),const utility::string_t &)' being compiled
`

If SPDLOG_WCHAR_TO_UTF8_SUPPORT  is not required in Windows, then what can we use to let WCHAR and CHAR work without problems ?

**Comment by gabime** [comment]
You need to prependthe formst string with L

spdlog::info(L”Hello”)n

**Comment by malidiab** [comment]
This is not related to the current issue , but I can't log wide strings unless I use the SPDLOG_WCHAR_TO_UTF8_SUPPORT'  , so for every line a conversion is being made from wide char to UTF8 before logging it to file. so I think there will be always a performance bottleneck for windows when using wide strings. 

```
#include <iostream>
#include <spdlog/\spdlog.h>

int main()
{
    std::cout << "Hello World!\n";
    spdlog::info("Welcome to spdlog!");

	//The line below is causing compilation problems. the only way to fix it is to use "SPDLOG_WCHAR_TO_UTF8_SUPPORT"
	spdlog::info(L"This Text cause compilation error");
}
```
The fix of the windows event log sink is to just use ::FormatMessageA for now. but we need to support using wide strings as well. 
it is very common to log Unicode user names , file names .. to event log in windows.

**Comment by iko1** [comment]
It should fix this issue: https://github.com/gabime/spdlog/pull/1760

**Comment by gabime** [comment]
fixed in #1760
